cpp/dolfinx/mesh/Topology.cpp

```diff
@@ -427,143 +441,151 @@ mesh::create_topology(MPI_Comm comm,
-        else
-          local_vertex_set.insert(v[j]);
-      }
-    }
+  int mpi_rank = MPI::rank(comm);
 
-    int mpi_rank = MPI::rank(comm);
+  // Make a list of all vertex indices whose ownership needs determining
+  std::vector<std::int64_t> unknown_indices(ghost_boundary_vertices.begin(),
+                                            ghost_boundary_vertices.end());
+  std::unordered_map<std::int64_t, std::vector<int>> global_to_procs
+      = compute_index_sharing(comm, unknown_indices);
 
-    // Make a list of all vertex indices whose ownership needs determining
-    std::vector<std::int64_t> unknown_indices(ghost_boundary_vertices.begin(),
-                                              ghost_boundary_vertices.end());
-    std::unordered_map<std::int64_t, std::vector<int>> global_to_procs
-        = compute_index_sharing(comm, unknown_indices);
+  // Number all indices which this process now owns
+  std::int32_t c = 0;
+  for (std::int64_t gi : local_vertex_set)
+  {
+    // Locally owned
+    auto [it_ignore, insert] = global_to_local_index.insert({gi, c++});
+    assert(insert);
+  }
 
-    // Number all indices which this process now owns
-    std::int32_t c = 0;
-    for (std::int64_t gi : local_vertex_set)
+  for (std::int64_t gi : ghost_boundary_vertices)
+  {
+    const auto it = global_to_procs.find(gi);
+    assert(it != global_to_procs.end());
+
+    // Shared and locally owned
+    if (it->second[0] == mpi_rank)
     {
-      // Locally owned
-      auto [it_ignore, insert] = global_to_local_index.insert({gi, c++});
-      assert(insert);
+      // Should already be in map, but needs index
+      auto it_gi = global_to_local_index.find(gi);
+      assert(it_gi != global_to_local_index.end());
+      assert(it_gi->second == -1);
+      it_gi->second = c++;
     }
-    for (std::int64_t gi : ghost_boundary_vertices)
+  }
+  std::int32_t nlocal = c;
+
+  // Get global offset for local indices
+  std::int64_t global_offset = dolfinx::MPI::global_offset(comm, nlocal, true);
+
+  // Find all vertex-sharing neighbours, and process-to-neighbour map
+  std::set<int> vertex_neighbours;
+  for (auto q : global_to_procs)
+    vertex_neighbours.insert(q.second.begin(), q.second.end());
+  vertex_neighbours.erase(mpi_rank);
+  std::vector<int> neighbours(vertex_neighbours.begin(),
+                              vertex_neighbours.end());
+  std::unordered_map<int, int> proc_to_neighbours;
+  for (std::size_t i = 0; i < neighbours.size(); ++i)
+    proc_to_neighbours.insert({neighbours[i], i});
+
+  // Communicate new global index to neighbours
+  MPI_Comm neighbour_comm;
+  MPI_Dist_graph_create_adjacent(comm, neighbours.size(), neighbours.data(),
+                                 MPI_UNWEIGHTED, neighbours.size(),
+                                 neighbours.data(), MPI_UNWEIGHTED,
+                                 MPI_INFO_NULL, false, &neighbour_comm);
+
+  std::vector<std::vector<std::int64_t>> send_pairs(neighbours.size());
```

Comment:
Would be good if we could avoid these data structures.

Reply:
I decided to keep this one. Precomputing the sizes and offsets requires traversing a map, which seems possibly just as costly. The size here is small, anyway, just neighbours.

Reply:
Ok, it's fine if it's just neighbours.